torch/csrc/Stream.cpp

```diff
@@ -0,0 +1,112 @@
+#include <pybind11/pybind11.h>
+#include <torch/csrc/Device.h>
+#include <torch/csrc/THP.h>
+#include <torch/csrc/utils/python_arg_parser.h>
+
+#include <structmember.h>
+
+PyTypeObject *THPStreamClass = nullptr;
+
+static PyObject* THPStream_pynew(
+  PyTypeObject *type, PyObject *args, PyObject *kwargs) {
+  HANDLE_TH_ERRORS
+  uint64_t cdata = 0;
+  static char *kwlist[] = {"_cdata", nullptr};
+  if (!PyArg_ParseTupleAndKeywords(
+    args, kwargs, "|K", kwlist, &cdata)) {
+    return nullptr;
+  }
+
+  THPObjectPtr ptr(type->tp_alloc(type, 0));
+  if (!ptr) {
+    return nullptr;
+  }
+
+  THPStream* self = (THPStream *)ptr.get();
+  self->cdata = cdata;
+  return (PyObject *)ptr.release();
+  END_HANDLE_TH_ERRORS
+}
+
+static PyObject * THPStream_get_device(THPStream *self, void *unused) {
+  HANDLE_TH_ERRORS
+  return THPDevice_New(c10::Stream::unpack(self->cdata).device());
+  END_HANDLE_TH_ERRORS
+}
+
+static PyObject * THPStream_eq(THPStream *self, THPStream *other) {
+  HANDLE_TH_ERRORS
+  return PyBool_FromLong(self->cdata == other->cdata);
+  END_HANDLE_TH_ERRORS
+}
+
+static struct PyMemberDef THPStream_members[] = {
+  {(char*)"_cdata",
+    T_ULONGLONG, offsetof(THPStream, cdata), READONLY, nullptr},
+  {nullptr}
+};
+
+static struct PyGetSetDef THPStream_properties[] = {
+  {"device", (getter)THPStream_get_device, nullptr, nullptr, nullptr},
+  {nullptr}
+};
+
+static PyMethodDef THPStream_methods[] = {
+  {(char*)"__eq__", (PyCFunction)THPStream_eq, METH_O, nullptr},
+  {nullptr}
+};
+
+PyTypeObject THPStreamType = {
+  PyVarObject_HEAD_INIT(&PyType_Type, 0)
```

Comment:
Huh, THCPStreamType previously was nullptr here. Is this the right way to do it?

Reply:
PyType_Type introduce more properties for CLASS type. Both nullptr and PyType_Type work. Right?